### ISSUE TITLE
[$1000] [HOLD for payment 2022-12-20] BUG: Offline requests are triggered after signing out

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#11850]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Go to new.expensify.com. 
2. Enter email. Click Continue. 
3. Enter password. Click Signin. 
5. Press command + option + J to open JS console. 
6. Click Network in JS console.
7. Turn off your internet connection. 
8. Click three different reports in LHN, and notice API requests are queued. 
9. Open settings page.
10. Click Sign out. 
11. Turn on your internet connection
12. Reload the newdot sign in page.
13. Notice that the requests taken while logged in are triggered in a loop and if you click on any of the requests and they all fail with session expired. 

## Expected Result:
No requests should be triggered

## Actual Result:
Requests taken while signed in are triggered

## Workaround:
unknown

## Platform:

Where is this issue occurring?

- Web
- iOS - Could not check API request in app
- Android - Could not check API request in app
- Desktop App
- Mobile Web

**Version Number:** 1.2.15-2
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/195892837-580e6245-7dda-4339-8ca8-db3d5a37ac4c.mov

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @luacmartins 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1665694797069499

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)
<img width="960" alt="mweb" src="https://user-images.githubusercontent.com/43996225/195892870-a4216e61-c6b6-46c6-a373-338e8eb343a5.png">
<img width="960" alt="Screenshot (245)" src="https://user-images.githubusercontent.com/43996225/195892875-289d8d76-5628-4dd0-91fb-7abb17c2083a.png">

## Upwork Automation - Do Not Edit
- Upwork Job URL: https://www.upwork.com/jobs/~017e848943cbf36ad1
- Upwork Job ID: 1599834031670763520
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01eb5d4aa7332964cc</li>
        <li>Upwork Job ID: 1604963994704998400</li>
        <li>Last Price Increase: 2022-12-19</li>
    </ul>
</details>

**Comment by joekaufmanexpensify** [comment]
Having trouble reproducing this. @luacmartins , do I need to have the JS console open prior to logging out? 

![2022-10-14_14-54-12 (1)](https://user-images.githubusercontent.com/54866469/195920551-302a68fa-0071-4f38-b0f1-c26db92001f4.gif)

**Comment by luacmartins** [comment]
@joekaufmanexpensify I just managed to reproduce again, but this time I had to refresh the page.

https://user-images.githubusercontent.com/22219519/196231973-3f16bb2b-4289-4650-8fd3-d8528eb8d9fb.mov

**Comment by mountiny** [comment]
I ma not sure how we would prevent this as logging out will remove the authtoken and make it stale so making the requests after cannot be authenticated (makes sense). What do you think Carlos

**Comment by luacmartins** [comment]
I think that we should clear the request queue so that these requests are not triggered at all~

**Comment by mountiny** [comment]
> I think that we should clear the request queue so that these requests are not triggered at all!

**Comment by joekaufmanexpensify** [comment]
Okay I was able to reproduce. Added some more detail to repro steps in OP. 

![2022-10-18_13-38-55 (1)](https://user-images.githubusercontent.com/54866469/196505067-f1014c73-c960-4257-a5e0-974a777b9184.gif)

**Comment by neil-marcellini** [comment]
I think we should keep this internal because it has to do with our network request queues / API.

If I send 4 messages in a chat while offline, sign out, refresh, and then go back online I see those requests fail. When I sign back in I have lost those messages, which is not good. It seems to me that we should save those requests and retry them when we log back in to that account. IDK if that's our plan, but it seems like a reason not to clear the request queue.

I also think we should be able to prevent trying these requests when we log out, like Carlos said. cc @marcaaron since he wrote a lot of this.

**Comment by isabelastisser** [comment]
Internal, unassigning myself.

**Comment by isabelastisser** [comment]
Internal, unassigning myself.

**Comment by michaelhaxhiu** [comment]
This actually should stay assigned to a member of the Bug Zero team. We are going to be holding onto `Internal` and `External` issues to sherpa them to completion.

**Comment by michaelhaxhiu** [comment]
Re-assigning Joe since he was initially assigned as the triager. But note for @isabelastisser in future GHs where you win the roulette spin!

**Comment by joekaufmanexpensify** [comment]
Discussing [next steps](https://expensify.slack.com/archives/C01SKUP7QR0/p1666296304223589).

**Comment by joekaufmanexpensify** [comment]
Updated repro steps to be even more clear. Requesting that an engineer add demolition label!

**Comment by joekaufmanexpensify** [comment]
Figured I'd test just to make sure I couldn't do this myself (since [we want Bug Zero](https://expensify.slack.com/archives/C01SKUP7QR0/p1666296503315919?thread_ts=1666296312.358889&cid=C01SKUP7QR0) to add demolition going forward).

**Comment by michaelhaxhiu** [comment]
@luacmartins can you apply the `demolition` label since you confirmed it was reproducible? In the future anyone will be able to add that label (https://github.com/Expensify/Expensify/issues/237144) but for now only Engineers can do it.

**Comment by MariaHCD** [comment]
Catching up here, looks like when the user signs out when offline we can either:
1. Clear the request queue. This would mean any offline requests would be lost and we would not retry a request only for it to fail with `session expired`
2. Keep the request queue and retry the requests when the user logs back in. I'd think this would a better UX. Not super familiar with the logic for the request queue but there must be a way we can use the latest authToken (instead of the expired one) to retry those requests?

cc: @marcaaron

**Comment by marcaaron** [comment]
I'm with @neil-marcellini that the expected behavior might be that we send the messages first, then sign you out after. But could also see the clearing the request queue option to prevent failing network requests (though it seems less desirable).

@MariaHCD That summary looks correct to me.

I don't agree with this being a `Demolition` issue. It's more of a `Planning` issue. We need to discuss how this should work at a fundamental level (here or in Slack) and once we land on that we can figure out how to proceed technically.

**Comment by joekaufmanexpensify** [comment]
Got it. Does there needing to be a discussion to identify the best path forward disqualify an issue from being demolition? I had [thought](https://stackoverflow.com/c/expensify/questions/8110) as long as it's a bug with clear, reproducible steps, that it would qualify. But LMK if that's not the case!

**Comment by marcaaron** [comment]
I don't think we need to have a discussion about the label. We could just leave the label on for now. Demolition issues are usually reproducible and have a clear fix - we assign someone automatically so they can get in there quickly and fix it. Since we need to do some planning to solve this it feels less like low hanging fruit.

**Comment by joekaufmanexpensify** [comment]
Okay, got it. I'm not too familiar with the different solutions, but happy to start a discussion about it in Slack to get the ball rolling!

**Comment by joekaufmanexpensify** [comment]
Posted discussion [here](https://expensify.slack.com/archives/C03TQ48KC/p1666967613458099).

**Comment by joekaufmanexpensify** [comment]
We seemed to have reached a consensus in Slack. Summarizing the discussion there before updating the issue.

**Comment by joekaufmanexpensify** [comment]
Noting from the [slack thread](https://expensify.slack.com/archives/C03TQ48KC/p1667328481839439?thread_ts=1666967613.458099&cid=C03TQ48KC) that we have consensus for how we should move forward with solving this. Specifically, when a user tries to sign out while offline, we should:

- Add a warning that alerts the user that any actions they took (messages, etc.) while offline will be lost if they move forward with signing out.
- Clear their request queue (meaning whatever action the user attempted while offline before signing out is lost).

Based on that discussion, we should be able to proceed with fixing this!

**Comment by MariaHCD** [comment]
Thanks, @joekaufmanexpensify! I'm interested in taking this on but will only be able to get to this next week.

**Comment by joekaufmanexpensify** [comment]
Sounds great! LMK if I can help.

**Comment by JmillsExpensify** [comment]
A [relavant discussion](https://expensify.slack.com/archives/C048ZLJ78P5/p1667496802620869?thread_ts=1667337295.224049&cid=C048ZLJ78P5) happened in Slack regarding how we would handle this scenario when 2FA is considered – more specifically, what happens if a user is offline, an admin enables 2FA on the domain and the user's session is invalidated.

**Comment by joekaufmanexpensify** [comment]
Added a few questions to the thread!

**Comment by joekaufmanexpensify** [comment]
All set to begin!

**Comment by MariaHCD** [comment]
WIP PR here: https://github.com/Expensify/App/pull/12596

**Comment by MariaHCD** [comment]
@joekaufmanexpensify @JmillsExpensify Could you confirm/improve the copy:

<img width="1105" alt="Screen Shot 2022-11-09 at 9 11 56 PM" src="https://user-images.githubusercontent.com/12268372/200895909-1a9a494e-f31d-498c-ad11-ee4b8d00f819.png">

**Comment by joekaufmanexpensify** [comment]
I feel like it could be nice to make it a bit lighter, perhaps something like `Watch out! You'll lose anything you did while offline if you sign out.` or something like that.

**Comment by joekaufmanexpensify** [comment]
I'm also curious, is the confirmation dialog in the middle of the screen like that standard? For example, when deleting a payment method, the confirmation dialog is completely within the right pane:

<img width="1439" alt="image" src="https://user-images.githubusercontent.com/54866469/200961432-4cbb2778-38cb-4d51-a87f-80c624809952.png">

I also feel like we might want to make the `Sign out` button red personally. As we want to discourage people from signing out here (I think).

**Comment by JmillsExpensify** [comment]
> I feel like it could be nice to make it a bit lighter, perhaps something like `Watch out! You'll lose anything you did while offline if you sign out.` or something like that.

Another alternative could be: `Are you sure? You'll lose any offline changes if you sign-out.`

**Comment by MariaHCD** [comment]
> when deleting a payment method, the confirmation dialog is completely within the right pane

@joekaufmanexpensify Isn't that a menu list vs a confirmation dialog?

I've tweaked the dialog to show the sign out button as red:

![Screen Shot 2022-11-10 at 1 10 12 PM](https://user-images.githubusercontent.com/12268372/201048657-dddcf0ef-82e4-49b0-ae66-90a1f12129c4.png)

> Another alternative could be: Are you sure? You'll lose any offline changes if you sign-out.

I'm onboard with `Are you sure? You'll lose any offline changes if you sign-out.`, if everyone agrees.

**Comment by MariaHCD** [comment]
Looking into the second aspect of this: `Clear their request queue (meaning whatever action the user attempted while offline before signing out is lost).`, I see that we already clear the queue upon sign out [here](https://github.com/Expensify/App/blob/599f1b9c806cfb612b74a3b9378f0d988d29b889/src/libs/actions/SignInRedirect.js#L64-L67) or I am missing something?

Also, I tried to reproduce the issue in the OP but I see that the `OpenReport` API commands are not being retried when coming back online. However, the `ReconnectApp` API command is retried several times and all of them fail with `Your session has expired. Please sign in again.`

https://user-images.githubusercontent.com/12268372/201064754-fa0171fd-a6ff-4d54-aa59-2516197a46d1.mov

cc: @luacmartins @marcaaron

**Comment by joekaufmanexpensify** [comment]
> Isn't that a menu list vs a confirmation dialog?

Yeah I guess it is more of a menu there. I was thinking of it as a confirmation as I associate clicking on the payment method as deleting it. More so, I am just curious if this is dialog is consistent with how we handle confirmation dialogs in other places in the app (which I am not too familiar with). If it is, then I think it makes sense!

>I'm onboard with Are you sure? You'll lose any offline changes if you sign-out., if everyone agrees.

I'm good with this!

**Comment by JmillsExpensify** [comment]
Fine with whatever on the copy side!

**Comment by joekaufmanexpensify** [comment]
PR is in draft.

**Comment by MariaHCD** [comment]
Still WIP

**Comment by joekaufmanexpensify** [comment]
PR still in draft.

**Comment by joekaufmanexpensify** [comment]
Same.

**Comment by joekaufmanexpensify** [comment]
@MariaHCD's PR is still WIP.

**Comment by joekaufmanexpensify** [comment]
Same.

**Comment by joekaufmanexpensify** [comment]
Not overdue.

**Comment by MariaHCD** [comment]
The [PR](https://github.com/Expensify/App/pull/12596) for showing the warning dialog is almost all set; I need to test it on all platforms and add the Spanish translation (asked [here](https://expensify.slack.com/archives/C21FRDWCV/p1669055295279659))

Going to look further into clearing the request queue because it seems like we do that already: https://github.com/Expensify/App/issues/11850#issuecomment-1310069976

**Comment by joekaufmanexpensify** [comment]
PR still WIP.

**Comment by MariaHCD** [comment]
Updated the PR https://github.com/Expensify/App/pull/12596 and it's out for review. In the meantime, going to look into  clearing the request queue

**Comment by MariaHCD** [comment]
The PR to show the warning dialog is merged: https://github.com/Expensify/App/pull/12596. Looking into clearing the request queue.

**Comment by Santhosh-Sellavel** [comment]
@MariaHCD Can you assign me this to track payment for review thanks!

**Comment by joekaufmanexpensify** [comment]
@MariaHCD There's another PR needed before this is issue is "complete", right?

**Comment by MariaHCD** [comment]
Yes, I think we'll need another PR to address the second part of this issue which was to clear the request queue if the user signs out when offline.

It does look like we already [clear the queue](https://github.com/Expensify/App/blob/599f1b9c806cfb612b74a3b9378f0d988d29b889/src/libs/actions/SignInRedirect.js#L64-L67) when a user signs out (https://github.com/Expensify/App/issues/11850#issuecomment-1310069976).

But we are triggering the `ReconnectApp` requests upon reconnection [here](https://github.com/Expensify/App/blob/main/src/libs/Navigation/AppNavigator/AuthScreens.js#L95). These requests all fail with a session expired error because we cleared the user's authToken. 

<img width="993" alt="Screenshot 2022-12-05 at 4 16 24 PM" src="https://user-images.githubusercontent.com/12268372/205686951-2ac36df4-b75f-42b5-8d63-083858702e57.png">

So I'm thinking we only call ReconnectApp if there is an authToken:

```diff
     componentDidMount() {
         NetworkConnection.listenForReconnect();
-        NetworkConnection.onReconnect(() => App.reconnectApp());
+        NetworkConnection.onReconnect(() => {
+            if (NetworkStore.getAuthToken()) {
+                App.reconnectApp();
+            }
+        });
         PusherConnectionManager.init();
```

@marcaaron @luacmartins Let me know what you think!

**Comment by Santhosh-Sellavel** [comment]
@joekaufmanexpensify  @MariaHCD  This [PR Review](https://github.com/Expensify/App/pull/12596) is due for payment today!

**Comment by luacmartins** [comment]
@MariaHCD IIRC we no longer use MainQueue after the API refactors and that's why we still trigger these requests. We now seem to be saving these requests to [PersistedRequests](https://github.com/Expensify/App/blob/18ae34f5b833f8561bed2bcf9f002eba4d6dafe5/src/libs/Network/SequentialQueue.js#L91), so we should be calling [PersistedRequests.clear()](https://github.com/Expensify/App/blob/18ae34f5b833f8561bed2bcf9f002eba4d6dafe5/src/libs/actions/PersistedRequests.js#L15-L18) when we log out.

**Comment by joekaufmanexpensify** [comment]
[Asked a quick question](https://expensify.slack.com/archives/C01SKUP7QR0/p1670264460554379) to make sure my understanding about payment for this issue is correct.

**Comment by joekaufmanexpensify** [comment]
In the interim, the only payment we need to make here is to @Santhosh-Sellavel, as this was reported internally, and we wrote the PR internally.

**Comment by joekaufmanexpensify** [comment]
Adding external label temporarily to create upwork job so we can pay Santhosh.

**Comment by Santhosh-Sellavel** [comment]
@joekaufmanexpensify Didn't receive any invites, so applied using this link. If you are not aware, you can hire me straight away instead of inviting me.

> https://www.upwork.com/jobs/~017e848943cbf36ad1

**Comment by joekaufmanexpensify** [comment]
Thanks for the heads up! Just sent you a contract.

**Comment by Santhosh-Sellavel** [comment]
Thanks, @joekaufmanexpensify Accepted the offer.

**Comment by joekaufmanexpensify** [comment]
Great, thanks! We're all set to issue payment here.

**Comment by joekaufmanexpensify** [comment]
[Discussed this ](https://expensify.slack.com/archives/C01SKUP7QR0/p1670344947493409?thread_ts=1670264460.554379&cid=C01SKUP7QR0)one more in Slack, and given there will be multiple PRs to fix this issue, we won't complete the bug zero checklist now. Instead, after the second PR has been merged, we'll complete it once then. With that in mind, updated the bug zero checklist above, and checked off the items.

**Comment by joekaufmanexpensify** [comment]
Next step for this issue is the second PR, being discussed [here. ](https://github.com/Expensify/App/issues/11850#issuecomment-1339662811)

**Comment by joekaufmanexpensify** [comment]
second PR is merged! 🎉

**Comment by joekaufmanexpensify** [comment]
QA is ongoing.

**Comment by joekaufmanexpensify** [comment]
Same.

**Comment by joekaufmanexpensify** [comment]
Same.

**Comment by joekaufmanexpensify** [comment]
Still undergoing QA

**Comment by MariaHCD** [comment]
Thinking about the BugZero checklist above, I'm thinking this bug was not caused by a specific PR(s). We probably just missed testing the scenario when a user logs out when offline? cc: @luacmartins

**Comment by luacmartins** [comment]
@MariaHCD I agree that this was an oversight within the API refactors doc, so any PRs that introduced this behavior did so in an expected way. I'm ok with just checking off those items in the checklist.

**Comment by MariaHCD** [comment]
Thanks, @luacmartins. Agreed!

**Comment by MariaHCD** [comment]
@joekaufmanexpensify For the last item in the BugZero checklist, I think it's still worth adding a test for the scenario when a user logs out when offline.

**Comment by joekaufmanexpensify** [comment]
Proposed regression test [here. ](https://expensify.slack.com/archives/C01SKUP7QR0/p1671487975421969)

**Comment by joekaufmanexpensify** [comment]
As far as payment goes, we reported and wrote this PR internally. All that is needed is payment to @Santhosh-Sellavel for review of the second PR.

**Comment by joekaufmanexpensify** [comment]
Going to temporarily make this external to create a new upwork job.

**Comment by joekaufmanexpensify** [comment]
@Santhosh-Sellavel offer sent for $1,000!

**Comment by joekaufmanexpensify** [comment]
Still discussing new regression test in Slack.

**Comment by joekaufmanexpensify** [comment]
@Santhosh-Sellavel bump on accepting the offer in Upwork so we can pay you. Thanks!

**Comment by joekaufmanexpensify** [comment]
Of course! Payment sent and contract ended!

**Comment by joekaufmanexpensify** [comment]
Closed job in upwork!

**Comment by joekaufmanexpensify** [comment]
Linked new regression test above in BZ checklist.

**Comment by joekaufmanexpensify** [comment]
Bug is fixed, all necessary payment is issued, and BZ checklist is complete. Closing. Thanks for the work on this one everyone!
